### PR TITLE
Improve error handling for rejected, blocked, and instructor users

### LINKED ISSUE #12982: Improve error handling for accepting/rejecting invitations
We want to improve the error handling if you try to accept/reject an invitation that can no longer have that happen. We can probably just refresh the page. We also want to improve the error handling for trying to invite/reject random ids. We also want to split rejection into a model function.

https://github.com/PrairieLearn/PrairieLearn/pull/12733#discussion_r2379838967

If we want to improve the user interface for this edge case, I would prefer to keep that in a separate PR.

_Originally posted by @reteps in https://github.com/PrairieLearn/PrairieLearn/pull/12733#discussion_r2373572233_

Also: https://github.com/PrairieLearn/PrairieLearn/pull/12733#discussion_r2377219031

### LINKED ISSUE #13118: Add improved error page for blocked students / cases where students can't self-enroll
We want a nicer error page for blocked students or cases where students can't self-enroll.

### LINKED ISSUE #13188: Prevent instructors from inviting other instructors
### Details

Currently you can invite instructors to a course instance. You shouldn't be able to do that.

### PR DESCRIPTION
**Pull request by reteps**
# Description

Closes #12982, closes #13188, closes #13118

Handles various weird transitions on the homepage. `(error)` means we want a friendly error message.

- [x] Invited -> Joined, double accept handled (no-op) **TEST CASE**
- [x] Invited -> Rejected, double reject handled (no-op) **TEST CASE**
- [x] Invited -> Joined, second tab Invited -> Rejected handled (error) **TEST CASE**
- [x] Invited -> Rejected, second tab Invited -> Joined handled (joins) **TEST CASE**
- [x] Joined -> Removed, double remove handled (no-op) **TEST CASE**
These edge cases should also now be handled, but I didn't test that they were. Worst case, a 403 will occur for these odd cases.
- [x] Joined -> Blocked, second tab Joined -> Removed (error)
- [x] Invited -> Joined -> Blocked, second tab Invite -> Joined (error)
- [x] Invited -> Joined -> Blocked, second tab Invite -> Rejected (error)
On invitations page:
- [x] user doesn't exist (error) **TEST CASE**
- [x] user is instructor (error) **TEST CASE**
- [x] user is blocked (should work, previously did not) **TEST CASE**
In enrollment locations:
- [x] handles blocked users on the enrollment page, the autoEnroll middleware and the enrollment code page with fancy page/message.
- [x] Handles the case where self enrollment isn't enabled on the homepage/other pages with fancy page/message.
- [x] Enrollment code form handles various cases

Adds a confirmation to deleting an invitation, as the other actions can be reverted, but that cannot be.

# Testing

I manually tested the instructor invitation changes. I manually tested that the autoEnroll middleware handles blocked users, and cases where self-enrollment is not enabled.

CleanShot 2025-10-29 at 12 30 47@2x
CleanShot 2025-10-29 at 12 57 46@2x
CleanShot 2025-10-29 at 12 59 18@2x
CleanShot 2025-10-29 at 12 31 35@2x
CleanShot 2025-10-29 at 12 45 45@2x
CleanShot 2025-10-29 at 12 56 40@2x

### CODE REVIEW

File: apps/prairielearn/src/models/enrollment.ts [hunk 1]
```diff
@@ -103,6 +103,8 @@ async function _enrollUserInCourseInstance({
 }): Promise<Enrollment> {
   assertHasRole(authzData, requestedRole);
 
+  assertEnrollmentStatus(lockedEnrollment, ['invited', 'removed', 'rejected']);
```
Comment (reteps) [inline apps/prairielearn/src/models/enrollment.ts:106]: This is the piece of code that corresponds to:

> If the user was 'blocked', this will throw an error.

File: apps/prairielearn/src/pages/enrollmentCodeRequired/enrollmentCodeRequired.tsx [hunk 2]
```diff
@@ -32,32 +29,37 @@ router.get(
     const existingEnrollment = await run(async () => {
       // We don't want to 403 instructors
       if (!hasRole(res.locals.authz_data, 'Student')) return null;
-      return await selectOptionalEnrollmentByUserId({
-        userId: res.locals.authn_user.user_id,
+      return await selectOptionalEnrollmentByUid({
```
Comment (reteps) [inline apps/prairielearn/src/pages/enrollmentCodeRequired/enrollmentCodeRequired.tsx:33]: This wouldn't pick up invited/rejected users correctly, this is a bug fix.

File: apps/prairielearn/src/pages/course_instance/lookup/lookupByEnrollmentCode.ts [hunk 3]
```diff
@@ -31,15 +35,55 @@ router.get(
...
+      return;
+    }
+
+    const authnUser: User = res.locals.authn_user;
+
+    const existingEnrollment = await selectOptionalEnrollmentByUserId({
+      userId: res.locals.authn_user.user_id,
+      courseInstance,
+      requestedRole: 'Student', // TODO: Should be 'System'
+      authzData: dangerousFullAuthzForTesting(),
```
Comment (reteps) [inline apps/prairielearn/src/pages/course_instance/lookup/lookupByEnrollmentCode.ts:49]: This is only used for fancy error messages. we don't care about feeding it the correct auth because of that fact. Other places this is used follow the same pattern.

File: apps/prairielearn/src/models/enrollment.test.ts [hunk 4]
```diff
@@ -138,13 +138,18 @@ describe('ensureEnrollment', () => {
...
-    });
+    try {
+      await ensureEnrollment({
+        courseInstance,
+        userId: user.user_id,
+        requestedRole: 'Student',
+        authzData: dangerousFullAuthzForTesting(),
+        actionDetail: 'implicit_joined',
+      });
+    } catch (error) {
```
Comment (reteps) [inline apps/prairielearn/src/models/enrollment.test.ts:150]: This is now detected upstream and a nicer error page is shown for this case, so we now 403 here.

File: apps/prairielearn/src/pages/instructorStudentDetail/components/OverviewCard.tsx [hunk 5]
```diff
@@ -167,6 +164,44 @@ export function OverviewCard({
...
+              <h2 class="modal-title h4" id="cancelInvitationModalLabel">
+                Confirm cancel invitation
+              </h2>
+              <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close" />
+            </div>
+            <div class="modal-body">
+              <p>
+                After an invitation is cancelled, the student will no longer appear in your course
+                and any progress they have made on assessments will be lost. They can still rejoin
+                the course by accepting an invitation or self-enrollment.
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/instructorStudentDetail/components/OverviewCard.tsx]: This feels pretty wordy. I am suggesting the following but we could probably find a middle ground somewhere too if this feels too short. 

```suggestion
                The student will no longer appear in your course and any progress they have made on assessments will be lost.
```

File: apps/prairielearn/src/pages/instructorStudentDetail/components/OverviewCard.tsx [hunk 6]
```diff
@@ -57,7 +57,7 @@ export function OverviewCard({
         )}
       </div>
       <div class="card-body">
-        <div class="d-flex align-items-center justify-content-between">
+        <div class="d-flex flex-column flex-md-row mb-2 mb-md-0 align-items-md-center justify-content-between">
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/instructorStudentDetail/components/OverviewCard.tsx:60]: Nice addition here. Looks good on narrow viewports.

File: apps/prairielearn/src/pages/course_instance/lookup/lookupByEnrollmentCode.ts [hunk 7]
```diff
@@ -31,15 +35,55 @@ router.get(
...
+    }
+
+    if (
+      courseInstance.self_enrollment_enabled_before_date &&
+      new Date() >= courseInstance.self_enrollment_enabled_before_date
+    ) {
+      res.status(403).json({
+        error: 'Self-enrollment is no longer allowed for this course',
       });
       return;
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/course_instance/lookup/lookupByEnrollmentCode.ts]: I don't think this is true. @reteps Can you confirm this?

**Review by mylesw27 — APPROVED** [review]
This looks good. I left one nit on the cancel modal wording below and left a follow up on the CodeRabbit comment. I don't see any blockers. Thoroughly tested locally and worked as expected.

File: apps/prairielearn/src/pages/instructorStudentDetail/components/OverviewCard.tsx (hunk 5 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorStudentDetail/components/OverviewCard.tsx]: that looks good!

File: apps/prairielearn/src/pages/course_instance/lookup/lookupByEnrollmentCode.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/src/pages/course_instance/lookup/lookupByEnrollmentCode.ts]: Nope, coderabbit is right. invited users should pass this check.

[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/src/pages/course_instance/lookup/lookupByEnrollmentCode.ts]: Fixed in [416eb44](https://github.com/PrairieLearn/PrairieLearn/pull/13202/commits/416eb4430c5561b7677bdad6a3f7224817ae4f34).

File: apps/prairielearn/src/models/enrollment.test.ts [hunk 8]
```diff
@@ -138,13 +138,18 @@ describe('ensureEnrollment', () => {
...
+      // The model function should throw an error if the user is blocked.
+      assert.equal(error.message, 'Access denied');
+    }
```
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/src/models/enrollment.test.ts:153]: Fixed in [416eb44](https://github.com/PrairieLearn/PrairieLearn/pull/13202/commits/416eb4430c5561b7677bdad6a3f7224817ae4f34)
